accounts/abi/bind/bind.go

```diff
@@ -160,11 +179,137 @@ func bindType(kind abi.Type) string {
 	}
 }
 
+// bindTypeJava converts a Solidity type to a Java one. Since there is no clear mapping
+// from all Solidity types to Java ones (e.g. uint17), those that cannot be exactly
+// mapped will use an upscaled type (e.g. BigDecimal).
+func bindTypeJava(kind abi.Type) string {
+	stringKind := kind.String()
+
+	switch {
+	case strings.HasPrefix(stringKind, "address"):
+		parts := regexp.MustCompile("address(\\[[0-9]*\\])?").FindStringSubmatch(stringKind)
+		if len(parts) != 2 {
+			return stringKind
+		}
+		if parts[1] == "" {
+			return fmt.Sprintf("Address")
+		}
+		return fmt.Sprintf("Addresses")
```

Comment:
So, the problem is that I can only get primitive types or simple structs across from Go to Java, because it has to go through a C bridge. This means that all fancy types such as Go slices (i.e. `[]common.Address` is forbidden. Since most of the API is auto-generated by `gomobile`, those parts will need to use custom `Addresses` type to represent Address slices. As to the binding, I could in theory make it here specific Java types and use some custom wrapping unwrapping inside the generated contract binding, but I though it would be confusing that the bindings uses native types whereas the remainder of the mobile API uses custom types.

My suggestion would be to use custom types everywhere if our hands are tied for now and if eventually we write a full Java canonically-typed wrapper too around the generated code, we can update abigen to use those. Imho currently it's more important to keep a simple and predictable API than to make it perfect Java from the first go.